extensions/lifecycle/tasks/LifecycleTask.js

```diff
@@ -59,78 +69,82 @@ class LifecycleTask extends BackbeatTask {
-            if (err) {
-                return done(err);
-            }
-            return done(null, { Contents: data.Contents });
-        });
+        ], done);
     }
 
     /**
      * Handles versioned objects (both enabled and suspended)
      * @param {object} bucketData - bucket data
-     * @param {object} params - params for AWS S3 `listObjectVersions`
+     * @param {array} bucketLCRules - array of bucket lifecycle rules
      * @param {Logger.newRequestLogger} log - logger object
      * @param {function} done - callback(error, data)
      * @return {undefined}
      */
-    _getObjectVersions(bucketData, params, log, done) {
-        const req = this.s3target.listObjectVersions(params);
-        attachReqUids(req, log);
-        async.waterfall([
-            next => req.send((err, data) => {
-                if (err) {
-                    log.error('error listing versioned bucket objects', {
-                        error: err,
-                        bucket: params.Bucket,
-                    });
-                    return next(err);
-                }
-                return next(null, data);
-            }),
-            (data, next) => {
-                if (data.IsTruncated) {
-                    // NOTE: this might only be needed if a LC rule defines
-                    //   a rule for processing NoncurrentVersion actions
-                    const entry = Object.assign({}, bucketData, {
-                        details: {
-                            keyMarker: data.NextKeyMarker,
-                            versionIdMarker: data.NextVersionIdMarker,
-                        },
-                    });
-                    return this.sendBucketEntry(entry, err => {
-                        if (!err) {
-                            log.debug(
-                                'sent kafka entry for bucket consumption', {
-                                    method: 'LifecycleTask._getObjectVersions',
-                                });
-                        }
-                        return next(null, data);
-                    });
-                }
-                return process.nextTick(() => next(null, data));
-            },
-        ], (err, data) => {
+    _getObjectVersions(bucketData, bucketLCRules, log, done) {
+        const paramDetails = {};
+
+        if (bucketData.details.versionIdMarker &&
+        bucketData.details.keyMarker) {
+            paramDetails.KeyMarker = bucketData.details.keyMarker;
+            paramDetails.VersionIdMarker = bucketData.details.versionIdMarker;
+        }
+
+        this._listVersions(bucketData, paramDetails, log, (err, data) => {
             if (err) {
+                // error already logged at source
                 return done(err);
             }
-            return done(null, {
-                Versions: data.Versions,
-                DeleteMarkers: data.DeleteMarkers,
-            });
+            const list = this._mergeSortVersionsLists(data.Versions,
```

Comment:
No big deal here, but variable names could be a bit more representative of what they are, like `allVersions`, `allVersionsWithStaleDate` etc. (just examples, feel free to choose appropriate names for the purpose).